code/iaas/auth-logic/src/main/java/io/cattle/platform/iaas/api/auth/dao/impl/AuthDaoImpl.java

```diff
@@ -129,70 +164,245 @@ public void updateAccount(Account account, String name, String kind, String exte
+                                update.add(memberNew);
+                                deleteIgnore.add(member);
+                            } else {
+                                createIgnore.add(memberNew);
+                                deleteIgnore.add(member);
+                            }
+                        }
+                    }
+                }
+                HashSet<Member> create = (HashSet) members.clone();
+                create.removeAll(update);
+                create.removeAll(createIgnore);
+                HashSet<Member> delete = (HashSet) otherPreviosMembers.clone();
+                delete.removeAll(deleteIgnore);
+
+                Condition allMembers = DSL.falseCondition();
+                for (Member member : delete) {
+                    allMembers = allMembers.or(PROJECT_MEMBER.EXTERNAL_ID.eq(member.getExternalId())
+                            .and(PROJECT_MEMBER.EXTERNAL_ID_TYPE.eq(member.getExternalIdType()))
+                            .and(PROJECT_MEMBER.PROJECT_ID.eq(project.getId())));
+                }
+                List<? extends ProjectMember> toDelete = create().selectFrom(PROJECT_MEMBER).where(allMembers).fetch();
+                for (ProjectMember member : toDelete) {
+                    objectProcessManager.executeStandardProcess(StandardProcess.DEACTIVATE, member, null);
+                    objectProcessManager.executeStandardProcess(StandardProcess.REMOVE, member, null);
+                }
+                List<ProjectMember> newMembers = new ArrayList<>();
+                for (Member member : create) {
+                    newMembers.add(createProjectMember(project, member));
+                }
+                for (Member member : update) {
+                    newMembers.add(updateMemberRole(project, member));
+                }
+                newMembers.addAll(getProjectMembersByMembers(createIgnore, project.getId()));
+                return newMembers;
+            }
+        });
     }
 
-    @Inject
-    public void setGenericResourceDao(GenericResourceDao resourceDao) {
-        this.resourceDao = resourceDao;
+    private List<? extends ProjectMember> getProjectMembersByMembers(Set<Member> members, long projectId) {
+        Condition allMembers = DSL.falseCondition();
+        for (Member member: members){
+            allMembers = allMembers.or(PROJECT_MEMBER.EXTERNAL_ID.eq(member.getExternalId())
+                    .and(PROJECT_MEMBER.EXTERNAL_ID_TYPE.eq(member.getExternalIdType()))
+                    .and(PROJECT_MEMBER.ROLE.in(member.getRole()))
+                    .and(PROJECT_MEMBER.PROJECT_ID.eq(projectId))
+                    .and(PROJECT_MEMBER.STATE.eq(CommonStatesConstants.ACTIVE))
+                    .and(PROJECT_MEMBER.REMOVED.isNull()));
+        }
+        return create().selectFrom(PROJECT_MEMBER).where(allMembers).fetch();
     }
 
+    private ProjectMember updateMemberRole(Account project, Member member) {
```

Comment:
James, is there a reason why updating the resource happens outside of projectMember resource lifecycle? Like for create/delete you actually execute the process, and it does all the state transitions for you. In this case the db record just gets updated w/o any state change indication. If you would update this resource using standard CRUD operation (resource.update), the state transition would be the following:

generic.resource.process.5=%s.update;inactive,active;inactive=updating-inactive,active=updating-active;updating-inactive=inactive,updating-active=active

(taken from code/iaas/logic-common/src/main/resources/META-INF/cattle/process/defaults.properties)